### DISCUSSION TITLE
Add support for Quarto

### DISCUSSION THREAD

**Discussion by ChenZhongPu** [LazyVim/LazyVim#1367] [Ideas]
Recently, I noticed an awesome scientific and technical publishing system [Quarto](https://quarto.org/docs/get-started/hello/neovim.html) , which also offers [quarto.nvim](https://github.com/quarto-dev/quarto-nvim).

I think it would be great LazyVim could add it into extras.

**Comment by danni-storm** [comment]
100%

**Comment by fredrmb** [comment]
It also looks like one of the main developers behind the [quarto-nvim](https://github.com/quarto-dev/quarto-nvim) plugin have created a [reference implementation](https://github.com/jmbuhr/lazyvim-starter-for-quarto) :)

**Reply by weberam2** [reply]
It looks like he took that down :(

**Reply by ArthurHowardMorris** [reply]
There is a new reference implementation [here](https://github.com/jmbuhr/quarto-nvim-kickstarter).  

I'm noting this for future readers, not because @weberam2 needed me to tell him :)

**Comment by weberam2** [comment]
I'd like to maybe open this topic back up
I'd also REALLY like to see Quarto as an Extra plugin

I'm trying to configure Quarto now with LazyVim, and am getting stuck on extending completion and lsp. I think I'm basically overriding the LazyVim config, which I don't want to be doing. I just want to add to it.

But I'm super confused how to make it work with `opts = function(_, opts)`
I tried reading https://lazyvim-ambitious-devs.phillips.codes/course/chapter-19/ but am pretty lost, not going to lie

Can anyone help? Or does anyone want to work together?

This is basically what I'd like to incorporate with LazyVim: https://github.com/jmbuhr/quarto-nvim-kickstarter

**Comment by weberam2** [comment]
Oh, I may have made some progress:

I made an `extend-completion.lua` file in my plugins folder:

```
return {
  {
    "hrsh7th/nvim-cmp",
    optional = true,
    dependencies = {
      "hrsh7th/cmp-calc",
      "hrsh7th/cmp-emoji",
      "kdheepak/cmp-latex-symbols",
      "jmbuhr/cmp-pandoc-references",
      "onsails/lspkind-nvim",
      "jmbuhr/otter.nvim",
    },
    opts = function(_, opts)
      opts.sources = opts.sources or {}
      table.insert(opts.sources, {
        name = { "nvim_lsp", "path", "calc", "emoji", "latex_symbols", "pandoc_references" },
      })
    end,
  },
}
```

Might not have everything he has in his completions, but I think it is mainly what we want...

**Reply by dotfrag** [reply]
I don't use quarto but that looks correct.

**Reply by dpetka2001** [reply]
`opts.sources` should be a list of items `{ name = "something" }`, not a single item of `name = { "something", "other", "anything" }`

**Reply by weberam2** [reply]
Like this?

```
opts = function(_, opts)
      opts.sources = opts.sources or {}
      table.insert(opts.sources, {
        name = { "nvim_lsp"},
        name = {"path"},
        name = {"calc"}
      })
    end,
```
??

**Reply by weberam2** [reply]
This is what ChatGPT recommended:

```
return {
  {
    "hrsh7th/nvim-cmp",
    optional = true,
    dependencies = {
      "hrsh7th/cmp-calc",
      "hrsh7th/cmp-emoji",
      "kdheepak/cmp-latex-symbols",
      "jmbuhr/cmp-pandoc-references",
      "onsails/lspkind-nvim",
      "jmbuhr/otter.nvim",
    },
    opts = function(_, opts)
      opts.sources = opts.sources or {}
      local new_sources = { "nvim_lsp", "path", "calc", "emoji", "latex_symbols", "pandoc_references" }
      for _, source in ipairs(new_sources) do
        table.insert(opts.sources, { name = source })
      end
    end,
  },
}
```

**Reply by dpetka2001** [reply]
Yes something like this. `table.insert()` can only insert a single value into a list-like table, thus you have to use a `for` loop to do what you want. Otherwise, you could also use `vim.list_extend()` without the `for` loop. 

Also `nvim_lsp` and `path` already exist, so you might want to skip them.

**Reply by weberam2** [reply]
Smart. Thank you!

**Comment by ArthurHowardMorris** [comment]
It seems that @jmbuhr is working on a LazyExtras addition for this:

Here is a link to the fork that he is working on: https://github.com/jmbuhr/LazyVim/tree/main/lua/lazyvim/plugins/extras 

Here is the quarto extra: https://github.com/jmbuhr/LazyVim/blob/main/lua/lazyvim/plugins/extras/lang/quarto.lua 

Here is the otter extra: https://github.com/jmbuhr/LazyVim/blob/main/lua/lazyvim/plugins/extras/lsp/otter.lua

**Reply by jmbuhr** [reply]
Almost forgot about that :see_no_evil: I opened the PR now (https://github.com/LazyVim/LazyVim/pull/5514)

**Reply by jmbuhr** [reply]
With the changes to `otter.nvim` I made some time ago, it requires essentially no configuration, so it's a very basic extra.

**Reply by ArthurHowardMorris** [reply]
Thanks! (For this and all your other quarto & nvim work)

I dropped the two extras into my `.config/nvim/lua/plugins/` and they have been working great. My use so far has not been anywhere near any potential edge cases.

**Reply by weberam2** [reply]
Hi @jmbuhr,

Your work on this has been amazing. I have been using your kickstarter for a while and love it. Adding the quarto and otter plugins from your extras to my LazyVim has been great too.

Some comments/ideas:
- I think maybe setting the keymap to <Leader>cqp for quarto preview, and <Leader>coa/cod for otter might make more sense, as LazyVim already has <Leader>q as 'quit', and <Leader>c as 'code'
- One thing that is missing with just having quarto and otter added is there is no molten/slurm. I'll take a look at your full Extras repo to see if I can figure out how you set it up
- Do you know if there is any benefit or conflicts with having the R extra setup as well?

Again, thank you so much for doing this. I know it's probably a huge headache trying to set these up for different distros

<3
